application_sandbox/external_memory_host/main.cpp

```diff
@@ -0,0 +1,498 @@
+  containers::unique_ptr<vulkan::VkFramebuffer> framebuffer_;
+  containers::unique_ptr<vulkan::DescriptorSet> cube_descriptor_set_;
+};
+
+template <typename T>
+class VkBufferExternalMemoryHost {
+ public:
+  VkBufferExternalMemoryHost(vulkan::VulkanApplication* application,
+                             logging::Logger* log, size_t num_images,
+                             VkBufferUsageFlagBits usage)
+      : device_(application->device()),
+        log_(log),
+        buffer_(VK_NULL_HANDLE, nullptr, &application->device()),
+        device_memory_(VK_NULL_HANDLE, nullptr, &application->device()),
+        aligned_data_size_(
+            vulkan::RoundUp(sizeof(T), vulkan::kMaxOffsetAlignment)) {
+    VkBuffer buffer;
+
+    VkBufferCreateInfo create_info = {
+        VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO,  // sType
+        nullptr,                               // pNext
+        0,                                     // flags
+        aligned_data_size_ * num_images,       // size
+        VK_BUFFER_USAGE_UNIFORM_BUFFER_BIT |
+            VK_BUFFER_USAGE_TRANSFER_DST_BIT,  // usage
+        VK_SHARING_MODE_EXCLUSIVE,
+        0,
+        nullptr};
+
+    LOG_ASSERT(
+        ==, log, VK_SUCCESS,
+        device_->vkCreateBuffer(device_, &create_info, nullptr, &buffer));
+
+    buffer_.initialize(buffer);
+
+    VkMemoryRequirements requirements;
+    device_->vkGetBufferMemoryRequirements(device_, buffer, &requirements);
+    allocation_size_ = vulkan::RoundUp(requirements.size, 4096);
+
+    VkPhysicalDeviceExternalMemoryHostPropertiesEXT
+        physical_device_external_memory_host_properties{
+            VK_STRUCTURE_TYPE_MEMORY_HOST_POINTER_PROPERTIES_EXT,
+            nullptr,
+            0,
+        };
+    VkPhysicalDeviceProperties2 physical_device_properties2{
+        VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PROPERTIES_2,
+        &physical_device_external_memory_host_properties,  // pNext;
+        {},                                                // properties
+    };
+    application->instance()->vkGetPhysicalDeviceProperties2KHR(
+        application->device().physical_device(), &physical_device_properties2);
+
+    if (physical_device_external_memory_host_properties
+            .minImportedHostPointerAlignment > 0) {
+      allocation_size_ = vulkan::RoundUp(
+          allocation_size_, physical_device_external_memory_host_properties
+                                .minImportedHostPointerAlignment);
+    }
+
+#ifdef _WIN32
+    host_pointer_ = VirtualAlloc(0, allocation_size_, MEM_RESERVE | MEM_COMMIT,
+                                 PAGE_READWRITE);
+#elif __linux__
+    host_pointer_ = mmap(nullptr, allocation_size_, PROT_READ | PROT_WRITE,
+                         MAP_PRIVATE | MAP_NORESERVE | MAP_ANONYMOUS, -1, 0);
+#endif
+    VkMemoryHostPointerPropertiesEXT host_pointer_properties = {
+        VK_STRUCTURE_TYPE_MEMORY_HOST_POINTER_PROPERTIES_EXT};
+    device_->vkGetMemoryHostPointerPropertiesEXT(
```

Comment:
Although we are allocating extra memory for minImportedHostPointerAlignment, we never actually try and align host_pointer_ to the required alignment, which  vkGetMemoryHostPointerPropertiesEXT requires.

Reply:
Done.